File: QuotesforMindandSoul/app/src/main/java/ro/danserboi/quotesformindandsoul/MyWorker.java
package ro.danserboi.quotesformindandsoul;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import java.util.List;
import java.util.Random;

import static android.content.Context.MODE_PRIVATE;

public class MyWorker extends Worker {

    public MyWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        List<Quote> allQuotes = QuoteRoomDatabase.getDatabase(getApplicationContext()).quoteDao().getAllQuotes();
        Random random = new Random();
        int randomQuoteIdx = random.nextInt(allQuotes.size());
        Quote quoteOfTheDay = allQuotes.get(randomQuoteIdx);
        // Add quote to SharedPreferences
        SharedPreferences mPreferences = getApplicationContext().getSharedPreferences(
                Utils.SHARED_PREF_FILE, MODE_PRIVATE);
        SharedPreferences.Editor preferencesEditor = mPreferences.edit();
        preferencesEditor.putString(Utils.DAILY_QUOTE_KEY, quoteOfTheDay.getWords());
        preferencesEditor.putString(Utils.DAILY_AUTHOR_KEY, quoteOfTheDay.getAuthor());
        preferencesEditor.putString(Utils.DAILY_CAT_KEY, quoteOfTheDay.getCategory());
        preferencesEditor.putBoolean(Utils.DAILY_BOOKMARK_KEY, quoteOfTheDay.getBookmarked());
        preferencesEditor.apply();
        showNotification(quoteOfTheDay.getAuthor(), quoteOfTheDay.getWords());
        return Result.success();
    }


    private void showNotification(String task, String desc) {
        // We open the QuoteOfTheDayActivity when we click on notification
        Intent intent = new Intent(getApplicationContext() , QuoteOfTheDayActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        PendingIntent resultPendingIntent = PendingIntent.getActivity(getApplicationContext(),
                0 /* Request code */, intent,
                PendingIntent.FLAG_UPDATE_CURRENT);

        NotificationManager manager = (NotificationManager) getApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);
        String channelName = getApplicationContext().getString(R.string.channel_name);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new
                    NotificationChannel(Utils.PRIMARY_CHANNEL_ID, channelName, NotificationManager.IMPORTANCE_HIGH);
            channel.enableLights(true);
            channel.setLightColor(Color.BLUE);
            channel.enableVibration(true);
            channel.setDescription(getApplicationContext().getString(R.string.notification_description));
            manager.createNotificationChannel(channel);
        }
        NotificationCompat.Builder builder = new NotificationCompat.Builder(getApplicationContext(), Utils.PRIMARY_CHANNEL_ID)
                .setContentIntent(resultPendingIntent)
                .setContentTitle(task)
                .setContentText(desc)
                .setSmallIcon(R.drawable.ic_quote)
                .setAutoCancel(true);
        // This is for older Android devices
        builder.setPriority(NotificationCompat.PRIORITY_HIGH)
                .setDefaults(NotificationCompat.DEFAULT_ALL);
        // We make the notification expandable
        builder.setStyle(new NotificationCompat.BigTextStyle()
                .setBigContentTitle(task)
                .bigText(desc));

        manager.notify(1, builder.build());
    }
}

File: QuotesforMindandSoul/app/src/main/java/ro/danserboi/quotesformindandsoul/MyWorker.kt
package ro.danserboi.quotesformindandsoul

import androidx.work.WorkerParameters
import android.content.Intent
import android.app.PendingIntent
import android.app.NotificationManager
import android.os.Build
import android.app.NotificationChannel
import android.content.Context
import android.content.SharedPreferences
import android.graphics.Color
import androidx.core.app.NotificationCompat
import androidx.work.Worker
import java.util.*

class MyWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {
    override fun doWork(): Result {
        val allQuotes: List<Quote> = QuoteRoomDatabase.getDatabase(applicationContext)?.quoteDao()!!.allQuotes
        val random = Random()
        val randomQuoteIdx: Int = random.nextInt(allQuotes.size)
        val quoteOfTheDay: Quote = allQuotes[randomQuoteIdx]
        // Add quote to SharedPreferences
        val mPreferences: SharedPreferences = applicationContext.getSharedPreferences(
                Utils.SHARED_PREF_FILE, Context.MODE_PRIVATE)
        val preferencesEditor: SharedPreferences.Editor = mPreferences.edit()
        preferencesEditor.putString(Utils.DAILY_QUOTE_KEY, quoteOfTheDay.words)
        preferencesEditor.putString(Utils.DAILY_AUTHOR_KEY, quoteOfTheDay.author)
        preferencesEditor.putString(Utils.DAILY_CAT_KEY, quoteOfTheDay.category)
        preferencesEditor.putBoolean(Utils.DAILY_BOOKMARK_KEY, quoteOfTheDay.isBookmarked)
        preferencesEditor.apply()
        showNotification(quoteOfTheDay.author!!, quoteOfTheDay.words)
        return Result.success()
    }

    private fun showNotification(task: String, desc: String) {
        // We open the QuoteOfTheDayActivity when we click on notification
        val intent = Intent(applicationContext, QuoteOfTheDayActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        val resultPendingIntent: PendingIntent = PendingIntent.getActivity(applicationContext,
                0 /* Request code */, intent,
                PendingIntent.FLAG_UPDATE_CURRENT)
        val manager: NotificationManager = applicationContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channelName: String = applicationContext.getString(R.string.channel_name)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(Utils.PRIMARY_CHANNEL_ID, channelName, NotificationManager.IMPORTANCE_HIGH)
            channel.enableLights(true)
            channel.lightColor = Color.BLUE
            channel.enableVibration(true)
            channel.description = applicationContext.getString(R.string.notification_description)
            manager.createNotificationChannel(channel)
        }
        val builder: NotificationCompat.Builder = NotificationCompat.Builder(applicationContext, Utils.PRIMARY_CHANNEL_ID)
                .setContentIntent(resultPendingIntent)
                .setContentTitle(task)
                .setContentText(desc)
                .setSmallIcon(R.drawable.ic_quote)
                .setAutoCancel(true)
        // This is for older Android devices
        builder.setPriority(NotificationCompat.PRIORITY_HIGH)
                .setDefaults(NotificationCompat.DEFAULT_ALL)
        // We make the notification expandable
        builder.setStyle(NotificationCompat.BigTextStyle()
                .setBigContentTitle(task)
                .bigText(desc))
        manager.notify(1, builder.build())
    }
}